### ISSUE TITLE
[NvimTree] Could not start the fs_event watcher for path /xxxx/xxxx : ENOSPC

### ISSUE THREAD

**Issue by dawaldow** [nvim-tree/nvim-tree.lua#2762]
### Description

**Describe the bug**
Each time I use CTRL + N, the message in the attached picture occurs for each file (or each directory).  

**To Reproduce**
1. open NVIM
2. CTRL+N
3. error occurs consistently and for each file in the directory
4. see picture

**Expected behavior**
No error should occur when activating CTRL+N.

**Screenshots**
![Screenshot from 2024-04-30 14-11-35](https://github.com/NvChad/NvChad/assets/57457955/bd5b2f6b-6efd-4ad8-bd0d-dd28cdc8ff10)
![Screenshot from 2024-04-30 14-16-33](https://github.com/NvChad/NvChad/assets/57457955/6e3bb28e-58a0-4c63-ac87-d94b1c96267f)

**Desktop (please complete the following information):**
 - Operating System: UBUNTU 24.04
 - Terminal: Gnome Terminal
 - Version of Neovim: 0.9.5

Using NVCHAD following basic instructions for Linux
![image](https://github.com/nvim-tree/nvim-tree.lua/assets/57457955/ad786414-aa46-4686-a244-23b948fb14b7)

No .git file in the ~/.config/nvim directory or my home directory.

### Neovim version

```text
NVIM v0.9.5
Build type: Release
LuaJIT 2.1.1703358377
```

### Operating system and version

Ubuntu 24.04 - Kernel 6.8.0-31

### Windows variant

N/A

### nvim-tree version

I don't seem to have the nvim-tree.lua file, but NvimTree works.

### Clean room replication

```lua
Using NVCHAD setup out of the box, no changes.
```

### Steps to reproduce

1. nvim .
2. CTRL+N
3. error appears for each file and/or directory

### Expected behavior

No error ... installed on my MAC with no errors.

### Actual behavior

See screenshots in the description.

**Comment by gegoune** [comment]
Please use steps and minimal configuration as described in https://github.com/nvim-tree/nvim-tree.lua/wiki/Troubleshooting#clean-room-replication. We cannot troubleshoot Nvim configuration distributions.

With that said, are you sure it's not a disk space issue?

**Comment by dawaldow** [comment]
No disk issues - 
![image](https://github.com/nvim-tree/nvim-tree.lua/assets/57457955/0f3a20fd-8f5f-4fbc-beeb-7001f9dddb97)

I was able to recreate the error/issue in the clean room environment with no changes to the lua file.  I attempted to setup the l"og =" information to create a log file, but it doesn't seem to work or I don't know where to put the "log =" information.  But with that said...

After reading the clean room link, at the bottom it talked about increasing the number of watchers.  I increased fs.inotify.max_user_watches and instances and the error went away.  I am not sure how this works, if it just delays the error as the number of watchers is consumed or if it doesn't reach the max (assuming number of files).  But I will watch and see how this behaves.

**Comment by alex-courtis** [comment]
> After reading the clean room link, at the bottom it talked about increasing the number of watchers. I increased fs.inotify.max_user_watches and instances and the error went away. I am not sure how this works, if it just delays the error as the number of watchers is consumed or if it doesn't reach the max (assuming number of files). But I will watch and see how this behaves.

That "checks out" although it does not explain `ENOSPC` at all; you should receive a reasonable message.

Guessing: quemu is using a lot of inotify, maybe OneDrive. Either way, the default limit is very low and needs to be increased by most users. My colleagues and I use intellij which almost immediately saturates inotify, hence we increase it.

**Comment by alex-courtis** [comment]
> I was able to recreate the error/issue in the clean room environment with no changes to the lua file. I attempted to setup the l"og =" information to create a log file, but it doesn't seem to work or I don't know where to put the "log =" information. But with that said...

The log will go in `${XDG_STATE_HOME}/nvim/nvim-tree.log`. You _should_ receive a startup message that notifies you of this file however there are bugs in nvim 0.9 that squash this; fixed in 0.10.

**Comment by alex-courtis** [comment]
@dawaldow any updates on this one?

This is a serious issue that we should address - logs will identify the problem.

**Comment by AndrewIzatt** [comment]
I just started getting this error and added 

```
fs.inotify.max_user_watches=524288
fs.inotify.max_user_instances=1024
```
to the bottom of my `/etc/sysctl.conf` file.

I then ran `sudo sysctl -p`, quit NVIM and restarted my terminal and it hasn't come back so far...

**Comment by alex-courtis** [comment]
> I just started getting this error and added
> 
> ```
> fs.inotify.max_user_watches=524288
> fs.inotify.max_user_instances=1024
> ```
> 
> to the bottom of my `/etc/sysctl.conf` file.
> 
> I then ran `sudo sysctl -p`, quit NVIM and restarted my terminal and it hasn't come back so far...

Thank you. I've updated the [wiki](https://github.com/nvim-tree/nvim-tree.lua/wiki/Troubleshooting#fsinotifymax_user_watches-exceeded) with `fs.inotify.max_user_instances = 1024` instead of 256.

**Comment by litoj** [comment]
@AndrewIzatt but this is useful only if you have inotify-tools installed, right?

I've tried uninstalling them, installing them, upping max instances to 1024 but it still keeps printing this message whenever I am in a really large directory (understandable) or in a git repo, as if it was trying to load the .git despite it being hidden.

It didn't use to be the case, but the last 6 months I am greeted with an orange screen whenever I open NvimTree 😅

**Comment by alex-courtis** [comment]
> but this is useful only if you have inotify-tools installed, right?

I'm not sure you need this; `fs.inotify` is a core part of the OS.

Each Linux distro seems to have slightly different ways to change these values. See [wiki: arch linux](https://github.com/nvim-tree/nvim-tree.lua/wiki/Troubleshooting#fsinotifymax_user_watches-exceeded)

> I've tried uninstalling them, installing them, upping max instances to 1024 but it still keeps printing this message whenever I am in a really large directory (understandable) or in a git repo, as if it was trying to load the .git despite it being hidden.

Perhaps the new value is not applying. You could confirm:

```
: ; sysctl -a | grep fs.inotify
fs.inotify.max_queued_events = 16384
fs.inotify.max_user_instances = 1024
fs.inotify.max_user_watches = 1048576
```

**Comment by litoj** [comment]
No, the value does get applied, but as I said, it tries to track .git and gitignored files despite having it set to not show them at all. Git is full of folders, so it makes sense it wouldn't suffice.

Now that I am thinking about it it looks like a completely separate issue 🤔

**Comment by alex-courtis** [comment]
> No, the value does get applied, but as I said, it tries to track .git and gitignored files despite having it set to not show them at all. Git is full of folders, so it makes sense it wouldn't suffice.
> 
> Now that I am thinking about it it looks like a completely separate issue 🤔

That sounds right. I'll be grateful if you raise a new issue with a full reproducer: a tar/zip that can demonstrate the problem will be necessary.
